### PR TITLE
feat: bounty max config

### LINKED ISSUE #245: `BOUNTY_HUNTER_MAX_OPENED_ISSUES` Configuration
`BOUNTY_HUNTER_MAX_OPENED_ISSUES` environment variable to override the maximum amount of issues that a bounty hunter can simultaneously work on. Set the default the value to `2`. This is intended to accommodate the following scenario:

- A bounty hunter begins working on a bounty. 
- The bounty hunter opens their pull request and waits for the review.
- The bounty hunter continues on the second issue.
- The bounty hunter handles the pull request review adjustments and closes, while opening the pull request for the second issue.

---

> Then set `BOUNTY_HUNTER_MAX_OPENED_ISSUES` to `2` or `3`. This way a bounty hunter has a plenty of space in terms of issues he wants to solve.
> 
> Example:
> 1. `BOUNTY_HUNTER_MAX_OPENED_ISSUES=2`
> 2. Bounty hunter creates a PR that solves issue 1
> 3. Bounty hunter starts issue 2
> 4. Project team merges PR for issue 1
> 5. Project team reopens issue 1 because bounty hunter has to redo it
> 6. Bounty hunter creates a PR that solves issue 2
> 
> At this stage a bounty hunter has 2 opened issues and has to redo issue 1 because he is not allowed to self assign more than `BOUNTY_HUNTER_MAX_OPENED_ISSUES=2` issues.
> 
> _Originally posted by @rndquu in https://github.com/ubiquity/bounty-bot/discussions/277#discussioncomment-5782932_

---

              We still need to create the feature to enforce this DevPool rule inside of the bot logic.

_Originally posted by @pavlovcik in https://github.com/ubiquity/bounty-bot/issues/134#issuecomment-1502229130_
            

P.S. When this task is ready we should also add the `BOUNTY_HUNTER_MAX_OPENED_ISSUES` to the [config generation](https://github.com/ubiquity/pay.ubq.fi/blob/main/static/secret.html) page

### PR DESCRIPTION
**Pull request by seprintour**
fixes #245

### CODE REVIEW

File: src/handlers/comment/handlers/assign.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
-import { addAssignees, addCommentToIssue, getCommentsOfIssue } from "../../../helpers";
+import { addAssignees, addCommentToIssue, getCommentsOfIssue, listIssuesForRepo } from "../../../helpers";
 import { getBotConfig, getBotContext, getLogger } from "../../../bindings";
 import { Payload, LabelItem, Comment, IssueType } from "../../../types";
 import { deadLinePrefix } from "../../shared";
 import { IssueCommentCommands } from "../commands";
 import { getWalletAddress } from "../../../adapters/supabase";
 
 export const assign = async (body: string) => {
+  let maxIssues = process.env.BOUNTY_HUNTER_MAX_OPENED_ISSUES;
```
Comment (rndquu) [inline src/handlers/comment/handlers/assign.ts:9]: This parameter `BOUNTY_HUNTER_MAX_OPENED_ISSUES` is supposed to be read from the bot's yml config https://github.com/ubiquity/bounty-bot/blob/development/src/bindings/config.ts#L21

Example config file: https://github.com/Alpha-Labs-Global/blockalizer/blob/master/.github/workflows/ubiquibot-config.yml

Pls refactor to read from the bot's config, not from an environment variable

**Review by rndquu — CHANGES_REQUESTED** [review]

File: src/handlers/comment/handlers/assign.ts (hunk 1 above)
Comment (pavlovcik) [reply src/handlers/comment/handlers/assign.ts:9]: That's my fault for not clarifying this in the bounty! But yes it is necessary to read from the yml file.

Comment (seprintour) [reply src/handlers/comment/handlers/assign.ts]: > Pls refactor to read from the bot's config, not from an environment variable

The latest commit is a refactor like you asked, now available from the config object

Thank you

File: src/handlers/comment/handlers/assign.ts [hunk 2]
```diff
@@ -22,7 +22,21 @@ export const assign = async (body: string) => {
     return;
   }
 
+  // List all the issues in the repository. It may include `pull_request`
+  const issues_opened = await listIssuesForRepo(IssueType.OPEN);
```
[the remark this answers is not in the document]
Comment (seprintour) [reply src/handlers/comment/handlers/assign.ts:26]: I'd create a function for getting a usernames assigned issue then, Github APIs doesn't give these data.. if not returning all the issues is crazy

**Comment by seprintour** [comment]
@rndquu resolved both here.. made a new function to loop and get assigned issues only, its now a helper

**Comment by seprintour** [comment]
@rndquu resolved both here.. made a new function to loop and get assigned issues only, its now a helper

**Review by rndquu — APPROVED** [review]

File: .github/ubiquibot-config.yml [hunk 3]
```diff
@@ -1,41 +1,42 @@
...
+    weight: 4
+    target: "Price: 400+ USD"
+  - name: "Priority: 4 (Emergency)"
+    weight: 5
+    target: "Price: 500+ USD"
 autoPayMode: true
-analyticsMode: true
\ No newline at end of file
+analyticsMode: true
+bountyHunterMax: 2
```
[the remark this answers is not in the document]
Comment (seprintour) [reply .github/ubiquibot-config.yml:42]: > I feel like we should eventually refactor all of the properties to be kebab case in this file.

Doesn't really go well, especially the fact that it'd often change from kebab to camelcase in other to be passed as a variable

[the remark this answers is not in the document]
Comment (seprintour) [reply .github/ubiquibot-config.yml:42]: We can make them more expressive while still in camel case, don't want to make it messy with the kebab to camelcase switch.. its better its camel-cased throughout

[the remark this answers is not in the document]
Comment (pavlovcik) [reply .github/ubiquibot-config.yml:42]: I'm following conventions I've seen with most GitHub actions yml files.

[the remark this answers is not in the document]
Comment (seprintour) [reply .github/ubiquibot-config.yml:42]: > I'm following conventions I've seen with most GitHub actions yml files.

I'd just switch the yml case then whilst keeping the camel-casing for the variables

[the remark this answers is not in the document]
Comment (rndquu) [reply .github/ubiquibot-config.yml]: Moving from camel case to kebab case breaks compatibility with existing configs. Also we would need to update the config generation part.

I agree that `maxConcurrentBounties` is more expressive but moving to kebab case is not part of this task.

[the remark this answers is not in the document]
Comment (seprintour) [reply .github/ubiquibot-config.yml]: > I agree that `maxConcurrentBounties` is more expressive but moving to kebab case is not part of this task.

Agreed, I could revert if @pavlovcik agrees to keep it camel whilst still expressive

[the remark this answers is not in the document]
Comment (pavlovcik) [reply .github/ubiquibot-config.yml]: We should make a new bounty for kebab-casing all of the property names.

**Review by pavlovcik — APPROVED** [review]
Looks like you renamed the property names. Thanks!

**Comment by seprintour** [comment]
> Looks like you renamed the property names. Thanks!

Yea, changed all of the yml variable to kebab so i think the issue is not needed anymore

**Comment by pavlovcik** [comment]
There's some filenames still camel cased but you don't have to worry about fixing those in this bounty.

File: src/types/config.ts [hunk 4]
```diff
@@ -40,6 +40,7 @@ export const UnassignConfigSchema = Type.Object({
 export const ModeSchema = Type.Object({
   autoPayMode: Type.Boolean(),
   analyticsMode: Type.Boolean(),
+  bountyHunterMax: Type.Number(),
```
[the remark this answers is not in the document]
Comment (seprintour) [reply src/types/config.ts:47]: > The original idea of ModeSchema is to have `boolean` values so remove `bountyHunterMax` from here. We can add a new `AssignSchema` schema property and put the `bountyHunterMax `in it

Fixed @0xcodercrane

**Comment by pavlovcik** [comment]
Why don't we just update all the configs to be kebab case and handle fully enabling kebab case support? It should take about an hour to do.

**Comment by rndquu** [comment]
> Why don't we just update all the configs to be kebab case and handle fully enabling kebab case support? It should take about an hour to do.

Overall we should maintain backwards compatibility. Right now we have only few repositories that use the bot so we can update all of the configs to kebab case but if we had 100 partners we would have to revert.

This [task](https://github.com/ubiquity/bounty-bot/issues/296) doesn't take into account the [config generation page](https://pay.ubq.fi/secret) where we should also enforce the kebab case

So we should not have rushed the kebab case feature as a part of this issue

I'll update the configs and the config generation page

**Comment by Steveantor** [comment]
> > Why don't we just update all the configs to be kebab case and handle fully enabling kebab case support? It should take about an hour to do.
> 
> Overall we should maintain backwards compatibility. Right now we have only few repositories that use the bot so we can update all of the configs to kebab case but if we had 100 partners we would have to revert.
> 
> This [task](https://github.com/ubiquity/bounty-bot/issues/296) doesn't take into account the [config generation page](https://pay.ubq.fi/secret) where we should also enforce the kebab case
> 
> So we should not have rushed the kebab case feature as a part of this issue
> 
> I'll update the configs and the config generation page

Yes, backwards compatibility is crucial
We can't rush such changes as we grow bigger.

**Comment by rndquu** [comment]
> `DefaultPriceConfig` is using _camelCase_

It doesn't break anything so we can fix it as a part of this https://github.com/ubiquity/bounty-bot/issues/296

**Comment by pavlovcik** [comment]
> > > Why don't we just update all the configs to be kebab case and handle fully enabling kebab case support? It should take about an hour to do.
> > 
> > 
> > Overall we should maintain backwards compatibility. Right now we have only few repositories that use the bot so we can update all of the configs to kebab case but if we had 100 partners we would have to revert.
> > This [task](https://github.com/ubiquity/bounty-bot/issues/296) doesn't take into account the [config generation page](https://pay.ubq.fi/secret) where we should also enforce the kebab case
> > So we should not have rushed the kebab case feature as a part of this issue
> > I'll update the configs and the config generation page
> 
> Yes, backwards compatibility is crucial We can't rush such changes as we grow bigger.

"move slowly and don't break things"

**Comment by rndquu** [comment]
> we should consider giving less weight to being so careful about it

This is not about being careful. This is about time we spend on a feature (refactoring to kebab case) that is not really necessary at the moment. 

We should concentrate on tasks that "move the needle":
- bot unit testing
- incentives for all commenting contributors
- gnosis safe support
- gnosis chain support

Some of the tasks we have (like kebab case refactoring) or this one https://github.com/ubiquity/ubiquity-dollar/issues/578 are nice to have but not really necessary at the moment so it's kind of "expensive" it terms of the time to waste resources for "nice to have" features instead of "must have" features.

**Pull request merged**